flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/RootArgumentTypeStrategy.java

```diff
@@ -0,0 +1,254 @@
+import static org.apache.flink.table.types.logical.utils.LogicalTypeCasts.supportsImplicitCast;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getLength;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getPrecision;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.getScale;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.hasFamily;
+import static org.apache.flink.table.types.logical.utils.LogicalTypeChecks.hasRoot;
+
+/**
+ * Strategy for an argument that corresponds to a given {@link LogicalTypeRoot} and nullability.
+ *
+ * <p>Implicit casts will be inserted if possible.
+ */
+@Internal
+public final class RootArgumentTypeStrategy implements ArgumentTypeStrategy {
+
+	private final LogicalTypeRoot expectedRoot;
+
+	private final boolean expectedNullability;
+
+	public RootArgumentTypeStrategy(LogicalTypeRoot expectedRoot, boolean expectedNullability) {
+		this.expectedRoot = Preconditions.checkNotNull(expectedRoot);
+		this.expectedNullability = expectedNullability;
+	}
+
+	@Override
+	public Optional<DataType> inferArgumentType(CallContext callContext, int argumentPos, boolean throwOnFailure) {
+		final DataType actualDataType = callContext.getArgumentDataTypes().get(argumentPos);
+		final LogicalType actualType = actualDataType.getLogicalType();
+
+		if (!expectedNullability && actualType.isNullable()) {
+			if (throwOnFailure) {
+				throw callContext.newValidationError(
+					"Unsupported argument type. Expected nullable type of root '%s' but actual type was '%s'.",
+					expectedRoot,
+					actualType);
+			}
+			return Optional.empty();
+		}
+
+		return findDataType(
+			callContext,
+			throwOnFailure,
+			actualDataType,
+			expectedRoot,
+			expectedNullability);
+	}
+
+	@Override
+	public Argument getExpectedArgument(FunctionDefinition functionDefinition, int argumentPos) {
+		// "< ... >" to indicate that this is not a type
+		if (!expectedNullability) {
+			return Argument.of("<" + expectedRoot + " NOT NULL>");
+		}
+		return Argument.of("<" + expectedRoot + ">");
+	}
+
+	@Override
+	public boolean equals(Object o) {
+		if (this == o) {
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()) {
+			return false;
+		}
+		RootArgumentTypeStrategy that = (RootArgumentTypeStrategy) o;
+		return expectedNullability == that.expectedNullability &&
+			expectedRoot == that.expectedRoot;
+	}
+
+	@Override
+	public int hashCode() {
+		return Objects.hash(expectedRoot, expectedNullability);
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * Finds a data type that is close to the given data type in terms of nullability and conversion
+	 * class but of the given logical root.
+	 *
+	 * <p>This method is shared with {@link FamilyArgumentTypeStrategy}.
+	 */
+	static Optional<DataType> findDataType(
```

Comment:
nit: Could we moved that to a helper class in the `strategies` package? It looks a bit counter-intuitive that an unrelated(`FamilyArgumentTypeStrategy` does not extend from this class) class uses this method.